### PR TITLE
✨ New PUBSUB - NATS JetStream support

### PR DESCRIPTION
**Pull request by mfreeman451**
## Pull Request Template

**Description:**

-   Adding support for NATS JetStream message broker
-   #584 
-   Highlight the motivation behind the changes and the expected benefits.

**Breaking Changes (if applicable):**

N/A

**Additional Information:**

-   Using NATS JetStream API
-   Include screenshots or code snippets (if necessary) to clarify the changes.

**Checklist:**

-   [x] I have formatted my code using  `goimport`  and  `golangci-lint`.
-   [ ] All new code is covered by unit tests.
-   [ ] This PR does not decrease the overall code coverage.
-   [x] I have reviewed the code comments and documentation for clarity.

**Thank you for your contribution!**

### CODE REVIEW

**Comment by aryanmehrotra** [comment]
Hey, could you please check this out 
https://github.com/gofr-dev/gofr/blob/development/pkg/gofr/datasource/README.md

It is mentioned to have a separate repo - but you can also create a separate package, the same way it is done for mongodb, cassandra...

this should be supported as an external package - such that users who are using it will only have it in the binary.

File: pkg/gofr/datasource/pubsub/nats/errors.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package nats
```
Comment (Umang01-hash) [inline pkg/gofr/datasource/pubsub/nats/errors.go:1]: All the errors defined here are keep exported, but i think we can make the errors unexported and only export the ones needed since they will not be used by user directly

File: pkg/gofr/datasource/pubsub/nats/health.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	"time"
+
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
+	h "gofr.dev/pkg/gofr/health"
+)
+
+const (
+	natsBackend            = "NATS"
+	jetstreamStatusOK      = "OK"
```
Comment (Umang01-hash) [inline pkg/gofr/datasource/pubsub/nats/health.go:14]: Use camel-case for `jetStream` in all the defined constants

File: pkg/gofr/datasource/pubsub/nats/health.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+			n.Logger.Debug("NATS health check: JetStream enabled")
+		}
+	} else if n.Js == nil {
+		n.Logger.Debug("NATS health check: JetStream not enabled")
+	}
+
+	return health
+}
+
+func getJetstreamStatus(ctx context.Context, js jetstream.JetStream) string {
```
Comment (Umang01-hash) [inline pkg/gofr/datasource/pubsub/nats/health.go:77]: `getJetstreamStatus` should use camel-case and can be renamed to `getJetStreamStatus`

File: pkg/gofr/datasource/pubsub/nats/nats_embedded.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import (
+	"time"
+
+	"github.com/nats-io/nats-server/v2/server"
+)
+
+const embeddedConnTimeout = 10 * time.Second
+
+// RunEmbeddedNATSServer starts a NATS server in embedded mode.
+func RunEmbeddedNATSServer() (*server.Server, error) {
```
Comment (Umang01-hash) [inline pkg/gofr/datasource/pubsub/nats/nats_embedded.go:12]: The `RunEmbeddedNATSServer` seems to be unused, please specify it's purpose ?

**Review by Umang01-hash — CHANGES_REQUESTED** [review]
- Some general comments:
1. Please only export the methods/functions/variables that are needed or defined in interface as exported, we can un-export rest of them.
2. We should use camel-case in variable naming and lower-case while logging.
3. We don't need to create a separate example for the NATS pubsub, we can include the example in documentation if needed.

**Comment by Umang01-hash** [comment]
@mfreeman451 also the tests are failing for various packages. Please fix it

File: pkg/gofr/container/container.go [hunk 5]
```diff
@@ -23,7 +27,6 @@ import (
 	"gofr.dev/pkg/gofr/metrics/exporters"
 	"gofr.dev/pkg/gofr/service"
 	"gofr.dev/pkg/gofr/version"
-	"gofr.dev/pkg/gofr/websocket"
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/container/container.go:26]: goimport changed the order is all, `	"gofr.dev/pkg/gofr/websocket"` is still included in the import list?

File: pkg/gofr/container/container.go [hunk 6]
```diff
@@ -67,74 +70,181 @@ func NewContainer(conf config.Config) *Container {
...
 	}
 
-	c.Debug("Container is being created")
+	return levelFetchConfig
+}
 
+func (c *Container) initializeMetrics() {
 	c.metricsManager = metrics.NewMetricsManager(exporters.Prometheus(c.GetAppName(), c.GetAppVersion()), c.Logger)
-
-	// Register framework metrics
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/container/container.go:96]: not sure why that was removed, restoring..

File: pkg/gofr/container/container.go [hunk 7]
```diff
@@ -67,74 +70,181 @@ func NewContainer(conf config.Config) *Container {
...
+		MaxWait:     c.getDuration(conf, "NATS_MAX_WAIT"),
+		BatchSize:   c.getIntConfig(conf, "NATS_BATCH_SIZE", 0),
+		MaxPullWait: c.getIntConfig(conf, "NATS_MAX_PULL_WAIT", 0),
+		Consumer:    conf.Get("NATS_CONSUMER"),
+		CredsFile:   conf.Get("NATS_CREDS_FILE"),
 	}
 
+	// Define the connection function
+	natsConnect := func(serverURL string, opts ...nats.Option) (n.ConnInterface, error) {
+		conn, err := nats.Connect(serverURL, opts...)
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/container/container.go:185]: updated

File: pkg/gofr/container/container.go [hunk 8]
```diff
@@ -67,74 +70,181 @@ func NewContainer(conf config.Config) *Container {
...
+
+func (w *natsConnWrapper) NatsConn() *nats.Conn {
+	return w.Conn
+}
+
+func (c *Container) initializeFile() {
 	c.File = file.New(c.Logger)
 }
 
+func (c *Container) getIntConfig(conf config.Config, key string, defaultValue int) int {
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/container/container.go:227]: updated

File: pkg/gofr/container/container.go [hunk 9]
```diff
@@ -67,74 +70,181 @@ func NewContainer(conf config.Config) *Container {
...
+		conn, err := nats.Connect(serverURL, opts...)
+		if err != nil {
+			return nil, err
+		}
+
+		return &natsConnWrapper{conn}, nil
+	}
+
+	// Define the JetStream creation function
+	jetstreamNew := func(nc *nats.Conn) (jetstream.JetStream, error) {
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/container/container.go:194]: removed

File: pkg/gofr/datasource/pubsub/interface.go [hunk 10]
```diff
@@ -19,7 +19,7 @@ type Subscriber interface {
 type Client interface {
 	Publisher
 	Subscriber
-	Health() datasource.Health
+	Health() health.Health
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/interface.go:22]: when i brought in the go.mod into the nats package it created a import cycle I couldn't resolve w/o moving the health stuff into its own package.

File: pkg/gofr/datasource/pubsub/nats/client.go [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+	Ctx     context.Context
+	Cancel  context.CancelFunc
+}
+
+type MessageHandler func(context.Context, jetstream.Msg) error
+
+// NATSClient represents a client for NATS JetStream operations.
+type NATSClient struct {
+	Conn          ConnInterface
+	Js            jetstream.JetStream
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go:48]: done

File: pkg/gofr/datasource/pubsub/nats/client.go [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+type Subscription struct {
+	Handler MessageHandler
+	Ctx     context.Context
+	Cancel  context.CancelFunc
+}
+
+type MessageHandler func(context.Context, jetstream.Msg) error
+
+// NATSClient represents a client for NATS JetStream operations.
+type NATSClient struct {
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go:46]: collides with our interface ```// Client represents the main NATS JetStream client.
type Client interface {
	Publish(ctx context.Context, subject string, message []byte) error
	Subscribe(ctx context.Context, subject string, handler MessageHandler) error
	Close(ctx context.Context) error
	DeleteStream(ctx context.Context, name string) error
	CreateStream(ctx context.Context, cfg StreamConfig) error
	CreateOrUpdateStream(ctx context.Context, cfg jetstream.StreamConfig) (jetstream.Stream, error)
	Health() health.Health
}```

File: pkg/gofr/datasource/pubsub/nats/client.go [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+}
+
+// DeleteTopic deletes a topic (stream) in NATS JetStream.
+func (n *NATSClient) DeleteTopic(ctx context.Context, name string) error {
+	n.Logger.Debugf("Deleting topic (stream) %s", name)
+
+	err := n.Js.DeleteStream(ctx, name)
+	if err != nil {
+		if errors.Is(err, jetstream.ErrStreamNotFound) {
+			n.Logger.Debugf("Stream %s not found, considering delete successful", name)
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go:71]: is this just for errors?

File: pkg/gofr/datasource/pubsub/nats/client.go [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+		return err
+	}
+
+	n.Logger.Debugf("Successfully deleted topic (stream) %s", name)
+
+	return nil
+}
+
+// NewNATSClient creates a new NATS client.
+func NewNATSClient(
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go]: cleaned up

File: pkg/gofr/datasource/pubsub/nats/client.go [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+
+			return nil // If the stream doesn't exist, we consider it a success
+		}
+
+		n.Logger.Errorf("failed to delete stream (topic) %s: %v", name, err)
+
+		return err
+	}
+
+	n.Logger.Debugf("Successfully deleted topic (stream) %s", name)
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go]: done

File: pkg/gofr/datasource/pubsub/nats/client.go [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+		return err
+	}
+
+	n.ProcessMessages(ctx, msgs, handler)
+
+	return msgs.Error()
+}
+
+// ProcessMessages processes messages from a consumer.
+func (n *NATSClient) ProcessMessages(ctx context.Context, msgs jetstream.MessageBatch, handler MessageHandler) {
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go:230]: done

File: pkg/gofr/datasource/pubsub/nats/client.go [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+	for msg := range msgs.Messages() {
+		if err := n.HandleMessage(ctx, msg, handler); err != nil {
+			n.Logger.Errorf("Error handling message: %v", err)
+		}
+	}
+}
+
+// HandleMessage handles a message from a consumer.
+func (n *NATSClient) HandleMessage(ctx context.Context, msg jetstream.Msg, handler MessageHandler) error {
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go:239]: need it exported to test

File: pkg/gofr/datasource/pubsub/nats/client.go [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+	if err := handler(ctx, msg); err != nil {
+		n.Logger.Errorf("Error handling message: %v", err)
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go]: done

File: pkg/gofr/datasource/pubsub/nats/client.go [hunk 19]
```diff
@@ -0,0 +1,348 @@
...
+		return n.NakMessage(msg)
+	}
+
+	return nil
+}
+
+// NakMessage naks a message from a consumer.
+func (n *NATSClient) NakMessage(msg jetstream.Msg) error {
+	if err := msg.Nak(); err != nil {
+		n.Logger.Errorf("Failed to NAK message: %v", err)
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go]: updated

File: pkg/gofr/datasource/pubsub/nats/client.go [hunk 20]
```diff
@@ -0,0 +1,348 @@
...
+		n.Logger.Errorf("failed to create or update stream: %v", err)
+
+		return nil, err
+	}
+
+	return stream, nil
+}
+
+// ValidateConfigs validates the configuration for NATS JetStream.
+func ValidateConfigs(conf *Config) error {
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go:327]: need it exported to test

File: pkg/gofr/datasource/pubsub/nats/errors.go (hunk 1 above)
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/errors.go:1]: needed for testing

File: pkg/gofr/datasource/pubsub/nats/health.go (hunk 2 above)
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/health.go:14]: done

File: pkg/gofr/datasource/pubsub/nats/health.go (hunk 3 above)
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/health.go:77]: done

File: pkg/gofr/datasource/pubsub/nats/nats_embedded.go (hunk 4 above)
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/nats_embedded.go]: removed, was used for testing

File: examples/using-subscriber-nats/readme.md [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+  -p 8222:8222 \
+  -p 6222:6222 \
+  -v /path/to/your/nats-server.conf:/etc/nats/nats-server.conf:ro \
+  -v /path/to/your/local/data/jetstream:/data/jetstream \
+  nats:latest -c /etc/nats/nats-server.conf
+```
+> Replace /path/to/your/nats-server.conf with the actual path to your NATS configuration file.
+> Replace /path/to/your/local/data/jetstream with the actual path to where you want to store JetStream data on your local machine.
+
+### **2. Build and Run the GoFr Application
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply examples/using-subscriber-nats/readme.md:25]: removed

File: pkg/gofr/container/container.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Umang01-hash) [reply pkg/gofr/container/container.go:26]: You need to place it with other `gofr.dev` dependencies at the last

File: pkg/gofr/datasource/pubsub/nats/client.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (Umang01-hash) [reply pkg/gofr/datasource/pubsub/nats/client.go]: No it's mainly for all log messages, we should follow `lower case` while starting to log

File: pkg/gofr/datasource/pubsub/nats/client.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (Umang01-hash) [reply pkg/gofr/datasource/pubsub/nats/client.go:327]: Which test are we talking about here?

File: pkg/gofr/datasource/pubsub/interface.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (Umang01-hash) [reply pkg/gofr/datasource/pubsub/interface.go:22]: And by changing it we are getting these errors:
Screenshot 2024-09-20 at 1 04 28 PM

File: pkg/gofr/datasource/pubsub/nats/client.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Umang01-hash) [reply pkg/gofr/datasource/pubsub/nats/client.go:46]: The interface name should be `NATS` and client should be named as Client

File: pkg/gofr/datasource/pubsub/nats/client.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (Umang01-hash) [reply pkg/gofr/datasource/pubsub/nats/client.go:239]: Which test are we talking about here?

**Comment by Umang01-hash** [comment]
@mfreeman451 Thank you for resolving the review comments. Now your PR contains failing tests and code. Please fix it.

**Comment by mfreeman451** [comment]
fixed the tests for some of the pubsub stuff failing, seeing some other stuff failing though, not sure if I am the cause of that? ```=== RUN   Test_SQLRetryConnectionInfoLog
could not connect with 'user' user to 'test' database at 'host:3201', error: dial tcp 103.224.182.246:3201: connect: operation timed out
    sql_test.go:283:
                Error Trace:    /Users/mfreeman/src/gofr/pkg/gofr/datasource/sql/sql_test.go:283
                Error:          "reading database configurations from config file\ngenerating database connection string for 'postgres'\nregistering sql dialect 'postgres' for traces\nconnecting to 'user' user to 'test' database at 'host:3201'\n" does not contain "retrying SQL database connection"
                Test:           Test_SQLRetryConnectionInfoLog
--- FAIL: Test_SQLRetryConnectionInfoLog (76.05s)
FAIL
FAIL    gofr.dev/pkg/gofr/datasource/sql        76.538s```

File: pkg/gofr/datasource/pubsub/nats/client.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go:327]: As soon as you change the package name of your test, for example, `package client`, to `package client_test`, you now are limited to only having access to exported functions from your test.

**Comment by Umang01-hash** [comment]
> fixed the tests for some of the pubsub stuff failing, seeing some other stuff failing though, not sure if I am the cause of that? `=== RUN Test_SQLRetryConnectionInfoLog could not connect with 'user' user to 'test' database at 'host:3201', error: dial tcp 103.224.182.246:3201: connect: operation timed out sql_test.go:283: Error Trace: /Users/mfreeman/src/gofr/pkg/gofr/datasource/sql/sql_test.go:283 Error: "reading database configurations from config file\ngenerating database connection string for 'postgres'\nregistering sql dialect 'postgres' for traces\nconnecting to 'user' user to 'test' database at 'host:3201'\n" does not contain "retrying SQL database connection" Test: Test_SQLRetryConnectionInfoLog --- FAIL: Test_SQLRetryConnectionInfoLog (76.05s) FAIL FAIL gofr.dev/pkg/gofr/datasource/sql 76.538s`

No @mfreeman451 you are not responsible for these test failures. They occur sometimes and can be corrected by re-running them

File: pkg/gofr/datasource/pubsub/nats/errors.go (hunk 1 above)
Comment (Umang01-hash) [reply pkg/gofr/datasource/pubsub/nats/errors.go:1]: @mfreeman451 if you want to write integration tests outside of the package you should do it in examples. Please specify where exactly are you using these exported errors for testing outside the package `nats`

File: pkg/gofr/datasource/pubsub/interface.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (Umang01-hash) [reply pkg/gofr/datasource/pubsub/interface.go:22]: The health functionality has been used by other datasources also (ex: `cassandra`, `dgraph` etc) but none of them felt the need to export heath to a different package. This will create unnecessary complications. You only need a method that can check the Health of the data source so that we can call it here:

```go
func (c *Container) Health(ctx context.Context) interface{} {
	var (
		healthMap = make(map[string]interface{})
		downCount int
	)

	const statusDown = "DOWN"

	if !isNil(c.SQL) {
		health := c.SQL.HealthCheck()
		if health.Status == statusDown {
			downCount++
		}

		healthMap["sql"] = health
	}

	if !isNil(c.Redis) {
		health := c.Redis.HealthCheck()
		if health.Status == statusDown {
			downCount++
		}

		healthMap["redis"] = health
	}

	if c.PubSub != nil {
		health := c.PubSub.Health()
		if health.Status == statusDown {
			downCount++
		}

		healthMap["pubsub"] = health
	}

	downCount += checkExternalDBHealth(ctx, c, healthMap)

	for name, svc := range c.Services {
		health := svc.HealthCheck(ctx)
		if health.Status == statusDown {
			downCount++
		}

		healthMap[name] = health
	}

	c.appHealth(healthMap, downCount)

	return healthMap
}
```

Feel free to take a reference from them. Also when checking out your branch on local to test the code i am getting unwanted go.mod errors regarding this health package:

```
gofr.dev/pkg/gofr/datasource/pubsub/nats imports
	gofr.dev/pkg/gofr/health: module gofr.dev@latest found (v1.21.0), but does not contain package gofr.dev/pkg/gofr/health
```


please try to avoid this

File: pkg/gofr/datasource/pubsub/nats/client.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (Umang01-hash) [reply pkg/gofr/datasource/pubsub/nats/client.go:46]: Please rename it accordingly

File: pkg/gofr/container/container.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Umang01-hash) [reply pkg/gofr/container/container.go:26]: Please revert this change

File: pkg/gofr/datasource/pubsub/nats/client.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go:327]: I don't know what you're asking for here, do you want to eliminate this test and reduce test coverage? If you change `client_test.go` from `nats_test` package to `nats` that is not go idiomatic and would break linters.

File: pkg/gofr/datasource/pubsub/nats/errors.go (hunk 1 above)
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/errors.go:1]: The contribution guidelines state that new integrations should have test coverage, this is a go idiomatic practice to place your tests in their own package.

File: pkg/gofr/container/mock_container.go [hunk 22]
```diff
@@ -7,11 +7,10 @@ import (
 	"testing"
 
 	"go.uber.org/mock/gomock"
-
-	"gofr.dev/pkg/gofr/datasource"
 	"gofr.dev/pkg/gofr/datasource/file"
 	"gofr.dev/pkg/gofr/datasource/pubsub"
 	"gofr.dev/pkg/gofr/datasource/sql"
+	h "gofr.dev/pkg/gofr/health"
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/container/mock_container.go:13]: 'health' collides with the package name.. what would you like it renamed to ?

[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/container/mock_container.go]: updated

**Comment by vipul-rawat** [comment]
Hey @mfreeman451, I think the issue is with line 38 in `go.mod` of `gofr`, you are trying to import the package which has not been published!

Also, you are introducing the packages
- github.com/nats-io/nats.go v1.37.0 // indirect
- github.com/nats-io/nkeys v0.4.7 // indirect
- github.com/nats-io/nuid v1.0.1 // indirect
But this is counterproductive as you've built the NATS datasource as a separate go module.

Not there are 2 things you could do - 
1. Remove the `go.mod` from the NATS implementation which would resolve most of your problems(speculative)
2. Remove the `pkg/gofr/datasource/nats` dependency from the container and build them as 2 separate modules which just work as the interface defined by gofr is implemented by the nats implementation (refer to Mongo, Cassandra, etc.. implementation)

**Comment by mfreeman451** [comment]
> Hey @mfreeman451, I think the issue is with line 38 in `go.mod` of `gofr`, you are trying to import the package which has not been published!
> 
> Also, you are introducing the packages
> 
> * github.com/nats-io/nats.go v1.37.0 // indirect
> * github.com/nats-io/nkeys v0.4.7 // indirect
> * github.com/nats-io/nuid v1.0.1 // indirect
>   But this is counterproductive as you've built the NATS datasource as a separate go module.
> 
> Not there are 2 things you could do -
> 
> 1. Remove the `go.mod` from the NATS implementation which would resolve most of your problems(speculative)
> 2. Remove the `pkg/gofr/datasource/nats` dependency from the container and build them as 2 separate modules which just work as the interface defined by gofr is implemented by the nats implementation (refer to Mongo, Cassandra, etc.. implementation)

The separate `go.mod` is a project requirement for new integrations, not mine. If you remove the nats dep from the container then you can't initialize the NATS pubsub.. I need to remove the replace directive from the go.mod..

**Comment by vipul-rawat** [comment]
> > Hey @mfreeman451, I think the issue is with line 38 in `go.mod` of `gofr`, you are trying to import the package which has not been published!
> > Also, you are introducing the packages
> > 
> > * github.com/nats-io/nats.go v1.37.0 // indirect
> > * github.com/nats-io/nkeys v0.4.7 // indirect
> > * github.com/nats-io/nuid v1.0.1 // indirect
> >   But this is counterproductive as you've built the NATS datasource as a separate go module.
> > 
> > Not there are 2 things you could do -
> > 
> > 1. Remove the `go.mod` from the NATS implementation which would resolve most of your problems(speculative)
> > 2. Remove the `pkg/gofr/datasource/nats` dependency from the container and build them as 2 separate modules which just work as the interface defined by gofr is implemented by the nats implementation (refer to Mongo, Cassandra, etc.. implementation)
> 
> The separate `go.mod` is a project requirement for new integrations, not mine. If you remove the nats dep from the container then you can't initialize the NATS pubsub..

@mfreeman451 I understand how we need to declare new modules for the new integrations(such as they are not a direct dependency in the applications that are not using it). You can check #1054, container package to see how @aryanmehrotra implemented Azure EventHub as an external pubsub dependency.

**Comment by mfreeman451** [comment]
Anything else left to do here?

**Comment by Umang01-hash** [comment]
@mfreeman451 your PR is passing all the tests but failing the code quality step.

Here's what i found when i looked into it:

Screenshot 2024-10-03 at 9 53 20 AM


Please resolve these issues.

**Comment by aryanmehrotra** [comment]
@mfreeman451 any update on the issues?

File: pkg/gofr/datasource/file/ftp/go.mod [hunk 23]
```diff
@@ -1,6 +1,8 @@
 module gofr.dev/pkg/gofr/datasource/file/ftp
 
-go 1.22
+go 1.22.3
+
+toolchain go1.23.1
```
Comment (Umang01-hash) [inline pkg/gofr/datasource/file/ftp/go.mod:5]: Can we revert this change in this PR as it is out of the scope of the PR.

File: pkg/gofr/datasource/file/sftp/go.mod [hunk 24]
```diff
@@ -1,6 +1,8 @@
 module gofr.dev/pkg/gofr/datasource/file/sftp
 
-go 1.22
+go 1.22.3
+
+toolchain go1.23.1
```
Comment (Umang01-hash) [inline pkg/gofr/datasource/file/sftp/go.mod:5]: Please revert this change as it is not related to this PR.

File: pkg/gofr/datasource/pubsub/nats/client_test.go [hunk 25]
```diff
@@ -0,0 +1,789 @@
...
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/nats-io/nats.go/jetstream"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/otel/trace/noop"
```
Comment (Umang01-hash) [inline pkg/gofr/datasource/pubsub/nats/client_test.go:12]: Please sort the imports here.

File: pkg/gofr/datasource/pubsub/nats/client_test.go [hunk 26]
```diff
@@ -0,0 +1,789 @@
...
+	"go.uber.org/mock/gomock"
+	"gofr.dev/pkg/gofr/datasource/pubsub"
+	"gofr.dev/pkg/gofr/logging"
+	"gofr.dev/pkg/gofr/testutil"
+)
+
+func TestValidateConfigs(*testing.T) {
```
Comment (Umang01-hash) [inline pkg/gofr/datasource/pubsub/nats/client_test.go:19]: The test `TestValidateConfigs` is incomplete. Please either complete it or delete it.

File: pkg/gofr/datasource/pubsub/nats/connection_manager.go [hunk 27]
```diff
@@ -0,0 +1,176 @@
+package nats
+
+import (
+	"context"
+	"time"
+
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
```
Comment (Umang01-hash) [inline pkg/gofr/datasource/pubsub/nats/connection_manager.go:8]: Please sort the imports here

File: pkg/gofr/datasource/pubsub/nats/subscription_manager_test.go [hunk 28]
```diff
@@ -0,0 +1,210 @@
...
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/nats-io/nats.go/jetstream"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
```
Comment (Umang01-hash) [inline pkg/gofr/datasource/pubsub/nats/subscription_manager_test.go:11]: Please sort the imports here

**Review by Umang01-hash — CHANGES_REQUESTED** [review]

File: pkg/gofr/datasource/file/ftp/go.mod (hunk 23 above)
Comment (mfreeman451) [reply pkg/gofr/datasource/file/ftp/go.mod:5]: not sure how that happened..

File: pkg/gofr/datasource/pubsub/nats/config.go [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+// New creates a new Client.
+func New(cfg *Config, logger pubsub.Logger) *PubSubWrapper {
+	if cfg == nil {
+		cfg = &Config{}
+	}
+
+	if cfg.BatchSize == 0 {
+		cfg.BatchSize = 100 // Default batch size
```
Comment (Umang01-hash) [inline pkg/gofr/datasource/pubsub/nats/config.go:36]: Let's make default batchSize a constant , in this way even if we want to change it someday it will be a single line change.

**Review by Umang01-hash — CHANGES_REQUESTED** [review]

File: pkg/gofr/datasource/pubsub/nats/stream_manager.go [hunk 30]
```diff
@@ -0,0 +1,91 @@
...
+	logger pubsub.Logger
+}
+
+func NewStreamManager(js jetstream.JetStream, logger pubsub.Logger) *StreamManager {
+	return &StreamManager{
+		js:     js,
+		logger: logger,
+	}
+}
+
```
Comment (Umang01-hash) [inline pkg/gofr/datasource/pubsub/nats/stream_manager.go:22]: It will be nice if we can add godoc comment for these methods as they are useful

Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/stream_manager.go:22]: added

File: pkg/gofr/datasource/pubsub/nats/committer.go [hunk 31]
```diff
@@ -0,0 +1,41 @@
+package nats
+
+import (
+	"log"
+
+	"github.com/nats-io/nats.go/jetstream"
+)
+
+// createTestCommitter is a helper function for tests to create a natsCommitter.
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/committer.go:9]: updated

File: pkg/gofr/datasource/pubsub/nats/config.go (hunk 29 above)
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/config.go:36]: updated

File: pkg/gofr/datasource/pubsub/nats/client.go [hunk 32]
```diff
@@ -0,0 +1,268 @@
...
+
+type messageHandler func(context.Context, jetstream.Msg) error
+
+// Connect establishes a connection to NATS and sets up JetStream.
+func (c *Client) Connect() error {
+	if err := c.validateAndPrepare(); err != nil {
+		return err
+	}
+
+	connManager := NewConnectionManager(c.Config, c.logger, c.natsConnector, c.jetStreamCreator)
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply pkg/gofr/datasource/pubsub/nats/client.go:40]: updated

File: docs/advanced-guide/using-publisher-subscriber/page.md [hunk 33]
```diff
@@ -175,6 +175,81 @@ docker run -d \
...
+| `NATS_MAX_WAIT` | Maximum wait time for batch requests | No | - | `5s` |
+| `NATS_BATCH_SIZE` | Maximum number of messages to pull in a single request | No | 0 | `100` |
+| `NATS_MAX_PULL_WAIT` | Maximum wait time for individual pull requests | No | 0 | `500ms` |
+| `NATS_CONSUMER` | Name of the NATS consumer | No | - | `my-consumer` |
+| `NATS_CREDS_FILE` | Path to the credentials file for authentication | No | - | `/path/to/creds.json` |
+
+#### Usage
+
+When subscribing or publishing using NATS JetStream, make sure to use the appropriate subject name that matches your stream configuration.
+For more information on setting up and using NATS JetStream, refer to the official NATS documentation.
```
[the remark this answers is not in the document]
Comment (mfreeman451) [reply docs/advanced-guide/using-publisher-subscriber/page.md:252]: added

**Comment by mfreeman451** [comment]
Removing context support from my integration is unacceptable, I am closing this Pull Request and will maintain my own fork of the integration and gofr that uses contexts on Connect() methods.
